flycontrols.js

```diff
@@ -0,0 +1,183 @@
+const Signal = require('signals')
+const mat4 = require('pex-math/mat4')
+const vec3 = require('pex-math/vec3')
+const quat = require('pex-math/quat')
+const euler = require('pex-math/euler')
+
+
+function Flycontrols (opts) {
+  this.type = 'Flycontrols'
+  this.enabled = true
+  this.changed = new Signal()
+  this.entity = null
+  this.dirty = false
+  this.camera =  false
+  this.currentYaw = [0,0,0,1]
+  this.currentPitch = [0,0,0,1]
+
+  this.isLocked = false
+
+  this.moveForward = false 
+  this.moveLeft = false
+  this.moveBackward = false
+  this.moveRight = false
+  this.moveUp = false
+  this.moveDown = false
+
+  const initialState = {
+    position: [0, 1, 0],
+    element: opts.element || document,
+    flySpeed : 0.1,
+    sensitivityX : 0.005,
+    sensitivityY : 0.005,
+    developerMode : false
+  }
+
+  this.set(initialState)
+  this.set(opts)
+}
+
+Flycontrols.prototype.init = function (entity) {
+  this.entity = entity
+  this.setup()
+}
+
+Flycontrols.prototype.set = function (opts) {
+  Object.assign(this, opts)
+  Object.keys(opts).forEach((prop) => this.changed.dispatch(prop))
+}
+
+Flycontrols.prototype.update = function () {
+  if(this.moveForward || 
+    this.moveBackward || 
+    this.moveLeft || 
+    this.moveRight ||
+    this.moveUp || 
+    this.moveDown ){
+    let dir = [0,0,0]; 
+
+    if(this.moveForward ) vec3.add(dir,[0,0,-1])
+    if(this.moveBackward) vec3.add(dir,[0,0,1])
+    if(this.moveLeft) vec3.add(dir,[-1,0,0])
+    if(this.moveRight) vec3.add(dir,[1,0,0])
+
+    let vec = vec3.multQuat(dir,this.entity.transform.rotation)
+
+    if(this.moveUp) vec3.add(vec,[0,1,0])
+    if(this.moveDown) vec3.add(vec,[0,-1,0])
+
+    vec3.normalize(vec)
+    vec3.scale(vec,this.flySpeed)
+    let newPos = vec3.add(this.entity.transform.position,vec)
+    this.entity.transform.set({ position : newPos})
+  }
+
+}
+
+Flycontrols.prototype.updateWindowSize = function () {
```

Comment:
What is this?

Reply:
An ancient relic of a simpler time. Its gone now.